Tribler/util.py

```diff
@@ -0,0 +1,11 @@
+"""
+This file contains various utility methods.
+"""
+from __future__ import absolute_import
+
+import sys
+
+if sys.version_info.major > 2:
+    cast_to_unicode_utf8 = lambda x: "".join([chr(c) for c in x]) if isinstance(x, bytes) else str(x)
+else:
+    cast_to_unicode_utf8 = lambda x: x.decode('utf-8')
```

Comment:
Just to double check, are you sure you want this?

In Python 3: put all bytes in a `bytes` into shorts (/`chr`s) of a unicode string
In Python 2: interpret all bytes in a `str` as utf-8 encoded data in a unicode string

Both will create a unicode string, but the Python 2 version will error out for all values which are not in the utf-8 character map (for example '\x80').

Reply:
> Both will create a unicode string, but the Python 2 version will error out for all values which are not in the utf-8 character map (for example '\x80').

I did not change the behavior for the Python 2 side and yes, it can error out. The code where I invoke `cast_to_unicode_utf8` _should_ be able to deal with this error.

Maybe we can make a `cast_to_unicode_safe` method in another PR that ignores characters that are not decodable?

Reply:
Also worthy of study are __six.ensure_binary(), six.ensure_str(), and six.ensure_text()__ which are in __master__ on GitHub but they have never been published in a release.
https://github.com/benjaminp/six/blob/master/six.py#L853-L908

Reply:
A new version of __six__ landed overnight that that officially ships those three functions in it.
* https://github.com/benjaminp/six/blob/master/CHANGES

Reply:
After some digging, I also hit some paydirt with this one:

```python
>>> u'\x80'.encode('raw_unicode_escape')
'\x80'
>>> '\x80'.decode('raw_unicode_escape')
u'\x80'
```

Reply:
$ python3 -c "'\x80'.decode('raw_unicode_escape')" --> AttributeError :-(

I have found that the easiest conversions are done with:
* __codecs.encode(b'x', 'hex')__
* __codecs.decode('x', 'hex')__
See: https://github.com/Tribler/tribler/pull/4100/files#diff-68ce5114e24c3b06aea5e0411154d52fR38

Reply:
@cclauss That was Python 2 syntax, here's Python 3 for the same:

```python
>>> codecs.encode(u'\x80', 'raw_unicode_escape')
'\x80'
>>> codecs.decode('\x80', 'raw_unicode_escape')
u'\x80'
```

Reply:
@qstokkink want me to change the `cast_to_unicode_utf8` to that?

Reply:
Are __cast_to_unicode_utf8()__ and the new __six.ensure_test()__ equivalent?

Reply:
@devos50 If this is the behavior you want:

```python
import codecs
a = b'abc'
print(type(a), a)
b = codecs.decode(a, 'raw_unicode_escape')
print(type(b), b)
```

```
<class 'bytes'> b'abc'
<class 'str'> abc
```

Personally I like this. I think I can get rid of some of my custom constructs in IPv8 now as well.

Reply:
@cclauss while I would say the ensure methods are helpful, the newest version of `six` is not available in the (stable) Debian repositories and we cannot use these methods safely across all linux distros. At one point, we might move to pip-based dependencies which provide us with newer versions of Linux packages.

Reply:
That would be good, we still have many files to port.